### PR TITLE
Better treatment for numerical errors

### LINKED ISSUE #313: Getting DomainError when fitting some models
A DomainError is raised when fitting some models (please see stack trace below). I have copied a model below that triggers this error (the model is meaningless; just the simplest I could generate to show the problem).

```
using StateSpaceModels

value = repeat([10.0], 40)
seasonality = 7
events = [0.0 for i in 1:40, j in 1:1]

ssm = StateSpaceModels.BasicStructuralExplanatory(value, seasonality, events)
StateSpaceModels.fit!(ssm)
```

The error is: 

DomainError with -3.5937804133234396e-10: log will only return a complex result if called with a complex argument. Try log(Complex(x)).

And the stack trace: 

[1] throw_complex_domainerror(f::Symbol, x::Float64)
    @ Base.Math .\math.jl:33
  [2] _log(x::Float64, base::Val{:ℯ}, func::Symbol)
    @ Base.Math .\special\log.jl:304
  [3] log
    @ .\special\log.jl:269 [inlined]
  [4] update_llk!
    @ C:\...\.julia\packages\StateSpaceModels\XjBwj\src\filters\univariate_kalman_filter.jl:198 [inlined]
  [5] update_kalman_state!(kalman_state::StateSpaceModels.UnivariateKalmanState{Float64}, y::Float64, Z::Vector{Float64}, T::Matrix{Float64}, H::Float64, R::Matrix{Float64}, Q::Matrix{Float64}, d::Float64, c::Vector{Float64}, skip_llk_instants::Int64, t::Int64)
    @ StateSpaceModels C:\...\.julia\packages\StateSpaceModels\XjBwj\src\filters\univariate_kalman_filter.jl:280
  [6] filter_recursions!(kalman_state::StateSpaceModels.UnivariateKalmanState{Float64}, sys::LinearUnivariateTimeVariant{Float64}, steadystate_tol::Float64, skip_llk_instants::Int64)
    @ StateSpaceModels C:\...\.julia\packages\StateSpaceModels\XjBwj\src\filters\univariate_kalman_filter.jl:339
  [7] optim_kalman_filter(sys::LinearUnivariateTimeVariant{Float64}, filter::UnivariateKalmanFilter{Float64})
    @ StateSpaceModels C:\...\.julia\packages\StateSpaceModels\XjBwj\src\filters\univariate_kalman_filter.jl:288
  [8] optim_loglike(model::BasicStructuralExplanatory, filter::UnivariateKalmanFilter{Float64}, unconstrained_hyperparameters::Vector{Float64})
    @ StateSpaceModels C:\...\.julia\packages\StateSpaceModels\XjBwj\src\kalman_filter_and_smoother.jl:15
  [9] #41
    @ C:\...\.julia\packages\StateSpaceModels\XjBwj\src\fit.jl:40 [inlined]
 [10] finite_difference_gradient!(df::Vector{Float64}, f::StateSpaceModels.var"#41#42"{UnivariateKalmanFilter{Float64}, BasicStructuralExplanatory}, x::Vector{Float64}, cache::FiniteDiff.GradientCache{Nothing, Nothing, Nothing, Vector{Float64}, Val{:central}(), Float64, Val{true}()}; relstep::Float64, absstep::Float64, dir::Bool)
    @ FiniteDiff C:\...\.julia\packages\FiniteDiff\KkXlb\src\gradients.jl:275
 [11] finite_difference_gradient!
    @ C:\...\.julia\packages\FiniteDiff\KkXlb\src\gradients.jl:224 [inlined]
 [12] (::NLSolversBase.var"#g!#44"{StateSpaceModels.var"#41#42"{UnivariateKalmanFilter{Float64}, BasicStructuralExplanatory}, FiniteDiff.GradientCache{Nothing, Nothing, Nothing, Vector{Float64}, Val{:central}(), Float64, Val{true}()}})(storage::Vector{Float64}, x::Vector{Float64})
    @ NLSolversBase C:\...\.julia\packages\NLSolversBase\cfJrN\src\objective_types\twicedifferentiable.jl:113
 [13] (::NLSolversBase.var"#fg!#45"{StateSpaceModels.var"#41#42"{UnivariateKalmanFilter{Float64}, BasicStructuralExplanatory}})(storage::Vector{Float64}, x::Vector{Float64})
    @ NLSolversBase C:\...\.julia\packages\NLSolversBase\cfJrN\src\objective_types\twicedifferentiable.jl:117
 [14] value_gradient!!(obj::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, x::Vector{Float64})
    @ NLSolversBase C:\...\.julia\packages\NLSolversBase\cfJrN\src\interface.jl:82
 [15] value_gradient!(obj::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, x::Vector{Float64})
    @ NLSolversBase C:\...\.julia\packages\NLSolversBase\cfJrN\src\interface.jl:69
 [16] value_gradient!(obj::Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}}, x::Vector{Float64})
    @ Optim C:\...\.julia\packages\Optim\6Lpjy\src\Manifolds.jl:50
 [17] (::LineSearches.var"#ϕdϕ#6"{Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}}, Vector{Float64}, Vector{Float64}, Vector{Float64}})(α::Float64)
    @ LineSearches C:\...\.julia\packages\LineSearches\Ki4c5\src\LineSearches.jl:84
 [18] secant2!(ϕdϕ::LineSearches.var"#ϕdϕ#6"{Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}}, Vector{Float64}, Vector{Float64}, Vector{Float64}}, alphas::Vector{Float64}, values::Vector{Float64}, slopes::Vector{Float64}, ia::Int64, ib::Int64, phi_lim::Float64, delta::Float64, sigma::Float64, display::Int64)
    @ LineSearches C:\...\.julia\packages\LineSearches\Ki4c5\src\hagerzhang.jl:368
 [19] (::LineSearches.HagerZhang{Float64, Base.RefValue{Bool}})(ϕ::Function, ϕdϕ::LineSearches.var"#ϕdϕ#6"{Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}}, Vector{Float64}, Vector{Float64}, Vector{Float64}}, c::Float64, phi_0::Float64, dphi_0::Float64)
    @ LineSearches C:\...\.julia\packages\LineSearches\Ki4c5\src\hagerzhang.jl:269
 [20] HagerZhang
    @ C:\...\.julia\packages\LineSearches\Ki4c5\src\hagerzhang.jl:101 [inlined]
 [21] perform_linesearch!(state::Optim.LBFGSState{Vector{Float64}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Vector{Float64}}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#19#21"}, d::Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}})
    @ Optim C:\...\.julia\packages\Optim\6Lpjy\src\utilities\perform_linesearch.jl:59
 [22] update_state!(d::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, state::Optim.LBFGSState{Vector{Float64}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Vector{Float64}}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#19#21"})
    @ Optim C:\...\.julia\packages\Optim\6Lpjy\src\multivariate\solvers\first_order\l_bfgs.jl:204
 [23] optimize(d::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, initial_x::Vector{Float64}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#19#21"}, options::Optim.Options{Float64, Nothing}, state::Optim.LBFGSState{Vector{Float64}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Vector{Float64}})
    @ Optim C:\...\.julia\packages\Optim\6Lpjy\src\multivariate\optimize\optimize.jl:54
 [24] optimize(d::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, initial_x::Vector{Float64}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#19#21"}, options::Optim.Options{Float64, Nothing})
    @ Optim C:\...\.julia\packages\Optim\6Lpjy\src\multivariate\optimize\optimize.jl:36
 [25] fit!(model::BasicStructuralExplanatory; filter::UnivariateKalmanFilter{Float64}, optimizer::Optimizer, save_hyperparameter_distribution::Bool)
    @ StateSpaceModels C:\...\.julia\packages\StateSpaceModels\XjBwj\src\fit.jl:42
 [26] fit!(model::BasicStructuralExplanatory)
    @ StateSpaceModels C:\...\.julia\packages\StateSpaceModels\XjBwj\src\fit.jl:35
 [27] top-level scope
    @ In[5]:8
 [28] eval
    @ .\boot.jl:373 [inlined]
 [29] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
    @ Base .\loading.jl:1196

### PR DESCRIPTION
**Pull request by guilhermebodin**
replace #327 
close #313

### CODE REVIEW

**Comment by raphaelsaavedra** [comment]
I think I actually prefer the approach in #327 because it doesn't split `logdet` into two functions, which is probably faster and doesn't lose the `logdet` advantages:

> Logarithm of matrix determinant. Equivalent to log(det(M)), but may provide increased accuracy and avoids overflow/underflow.

I'm not a big fan of using `try/catch` but in this case it seems fine?

**Comment by raphaelsaavedra** [comment]
Thanks for taking care of this btw :)

**Comment by guilhermebodin** [comment]
I did the benchmarks

```julia

using BenchmarkTools

function scalar_with_try_catch(n::Int, F::Float64, v::Float64)
    llk = 0.0
    HALF_LOG_2_PI = 0.5 * log(2*pi)
    for i in 1:n
        try 
            llk -= (
                HALF_LOG_2_PI + 0.5 * (log(F) + v^2 / F)
            )
        catch
            error("Numerical error, F is negative: $F")
        end
    end
    return llk
end

function scalar_with_check(n::Int, F::Float64, v::Float64)
    llk = 0.0
    HALF_LOG_2_PI = 0.5 * log(2*pi)
    for i in 1:n
        if F < 0
            error("Numerical error, F is negative: $F")
        end
        llk -= (
            HALF_LOG_2_PI + 0.5 * (log(F) + v^2 / F)
        )
    end
    return llk
end

n = 1000
F = 1.0
v = 1.0

julia> @btime scalar_with_try_catch($n, $F, $v)
  23.400 μs (0 allocations: 0 bytes)
-1418.9385332046932

julia> @btime scalar_with_check($n, $F, $v)
  5.714 μs (0 allocations: 0 bytes)
-1418.9385332046932

function vetorial_with_try_catch(n::Int, F::Matrix{Float64}, v::Vector{Float64})
    llk = 0.0
    HALF_LOG_2_PI = 0.5 * log(2*pi)
    for i in 1:n
        try 
            llk -=
                HALF_LOG_2_PI + 0.5 * (logdet(F) +
                v' * inv(F) * v)
        catch
            error("Numerical error, F is negative: $(F[i])")
        end
    end
    return llk
end

function vetorial_with_check(n::Int, F::Matrix{Float64}, v::Vector{Float64})
    llk = 0.0
    HALF_LOG_2_PI = 0.5 * log(2*pi)
    for i in 1:n
        detF = det(F)
        if detF < 0
            error("Numerical error, F is negative: $(F[i])")
        end
        llk -=
            HALF_LOG_2_PI + 0.5 * (log(detF) +
            v' * inv(F) * v)
    end
    return llk
end

function vetorial_with_check_and_logdet(n::Int, F::Matrix{Float64}, v::Vector{Float64})
    llk = 0.0
    HALF_LOG_2_PI = 0.5 * log(2*pi)
    for i in 1:n
        detF = det(F)
        if detF < 0
            error("Numerical error, F is negative: $(F[i])")
        end
        llk -=
            HALF_LOG_2_PI + 0.5 * (logdet(F) +
            v' * inv(F) * v)
    end
    return llk
end

n = 1000
F = Matrix{Float64}(I, 3, 3)
v = ones(3)

julia> GC.gc()

julia> GC.gc()

julia> @btime vetorial_with_try_catch($n, $F, $v)
  3.747 ms (4000 allocations: 218.75 KiB)
-2418.9385332047

julia> GC.gc()

julia> GC.gc()

julia> @btime vetorial_with_check($n, $F, $v) # might have been a little unlucky in this one
  8.012 ms (6000 allocations: 296.88 KiB)
-2418.9385332047

julia> GC.gc()

julia> GC.gc()

julia> @btime vetorial_with_check_and_logdet($n, $F, $v)
  6.133 ms (6000 allocations: 296.88 KiB)
-2418.9385332047
```

In the scalar case it makes more difference, not using the try catch, in the vetorial case my approach actually hurts performance.

I found this post https://discourse.julialang.org/t/try-catch-statement-really-slow-even-if-catch-is-never-executed/91744/6 revealing that each try means a few nanoseconds.

I am going to change it bakc to try catch statements

**Comment by raphaelsaavedra** [comment]
What about moving the try block outside the for loop? It doesn't matter in which step it fails, it only matters that it fails:

```
function scalar_with_try_catch(n::Int, F::Float64, v::Float64)
    llk = 0.0
    HALF_LOG_2_PI = 0.5 * log(2*pi)
    try
        for i in 1:n
            llk -= (
                HALF_LOG_2_PI + 0.5 * (log(F) + v^2 / F)
            )
        end
        catch
            error("Numerical error, F is negative: $F")
        end
    return llk
end
```

**Comment by guilhermebodin** [comment]
This was a test to mimic the package behavior, realistically we would have to put the try-catch blocks in the filter recursions. 
```julia
function filter_recursions!(
    kalman_state::MultivariateKalmanState{Fl},
    sys::LinearMultivariateTimeInvariant,
    steadystate_tol::Fl,
    skip_llk_instants::Int,
) where Fl
    RQR = sys.R * sys.Q * sys.R'
    @inbounds for t in 1:size(sys.y, 1)
        update_kalman_state!(
            kalman_state,
            sys.y[t, :],
            sys.Z,
            sys.T,
            sys.H,
            RQR,
            sys.d,
            sys.c,
            skip_llk_instants,
            steadystate_tol,
            t,
        )
    end
    return kalman_state.llk
end
```

It is also a viable option.

File: src/filters/regression_kalman_filter.jl [hunk 1]
```diff
@@ -58,9 +58,14 @@ function regression_update_v!(
...
-    kalman_state.llk -= (
-        HALF_LOG_2_PI + 0.5 * (log(kalman_state.F) + kalman_state.v^2 / kalman_state.F)
-    )
+    try 
+        kalman_state.llk -= (
+            HALF_LOG_2_PI + 0.5 * (log(kalman_state.F) + kalman_state.v^2 / kalman_state.F)
+        )
+    catch
+        @error("Numerical error in the log-likelihood calculation. F = $(kalman_state.F), v = $(kalman_state.v). F can only be positive.")
+        rethrow()
```
Comment (raphaelsaavedra) [inline src/filters/regression_kalman_filter.jl:67]: Wondering if this is any different than just

```
catch e
     @error("Numerical error... :, $e)
```

But looks good to me!

**Review by raphaelsaavedra — APPROVED** [review]

File: src/filters/regression_kalman_filter.jl (hunk 1 above)
Comment (guilhermebodin) [reply src/filters/regression_kalman_filter.jl:67]: this one does not stop the program, only logs to screen
```julia
julia> try; log(-10); catch; @error("oi"); rethrow() end
┌ Error: oi
└ @ Main REPL[102]:1
ERROR: DomainError with -10.0:
log was called with a negative real argument but will only return a complex result if called with a complex argument. Try log(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(f::Symbol, x::Float64)
   @ Base.Math .\math.jl:33
 [2] _log
   @ .\special\log.jl:295 [inlined]
 [3] log(x::Float64)
   @ Base.Math .\special\log.jl:261
 [4] log(x::Int64)
   @ Base.Math .\math.jl:1531
 [5] top-level scope
   @ REPL[102]:1

julia> try; log(-10); catch ex; @error("oi", ex) end
┌ Error: oi
│   ex =
│    DomainError with -10.0:
│    log was called with a negative real argument but will only return a complex result if called with a complex argument. Try log(Complex(x)).
└ @ Main REPL[103]:1
```

**Comment by guilhermebodin** [comment]
@raphaelsaavedra I moved the try-catch blocks to filter recursions. I think this is the best way to go. If there is any error, it will report the entire Kalman state and tell where the error happened.

File: src/filters/regression_kalman_filter.jl (hunk 1 above)
Comment (raphaelsaavedra) [reply src/filters/regression_kalman_filter.jl]: Ah, it's been a while since I programmed in julia... forgot `@error` just logs! Thanks

**Pull request merged**